text/1026-net-module.rst

```diff
@@ -0,0 +1,209 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-08-14
+    Authors: Scott Trinh <scott@edgedb.com>
+
+=============================
+RFC: EdgeDB Networking Module
+=============================
+
+This RFC outlines the addition of a networking module to EdgeDB’s
+standard library, providing functionality for making HTTP requests and
+sending SMTP messages.
+
+Motivation
+==========
+
+The EdgeDB Auth extension (ext::auth) currently sends emails directly to
+configured SMTP services at certain times in various authentication
+ceremonies and workflows:
+
+-  Sign up with email (sends email verification email)
+-  Password reset (sends password reset email)
+-  Magic link sign up (sends magic link email)
+-  Magic link sign in (sends magic link email)
+
+To allow full customization of emails and provide hooks into
+authentication events generally, we need to support sending webhook
+requests to the application directly. This would enable applications to
+trigger their own logic for sending fully customized and
+internationalized emails, SMS, log analytics, update usage billing
+records, etc.
+
+Specification
+=============
+
+Module Name
+-----------
+
+The new module will be named ``net``.
+
+Utility types
+-------------
+
+Since requests are sent asynchronously, we need a way to track the status of the
+request. This is done by using a ``net::Task`` type to represent a request and
+its current state. Each protocol will have its own ``net::Task`` concrete type.
+
+.. code-block:: edgeql
+
+  scalar type net::TaskState extending std::enums<Pending, InProgress, Complete, Failed>;
+
+  abstract type net::Task {
+    required state: net::TaskState;
+    required created_at: datetime;
+  }
```

Comment:
I think we also probably want a `code: int16` and/or `reason: str` to store issues with the actual networking.

Reply:
Ah I was thinking about something like:

```
  abstract type net::Task {
    required state: net::TaskState;
    required created_at: datetime;
    multi failed_results: net::Result;
  }

  type net::Result {
    required complete_at: datetime;
    required code: net::ErrorCode;
    required message: str;
    # maybe also link to responses (if received one) in subtypes
  };
```

So that the whole retrying process is auditable. Again, this is for the RFC not the MVP.

Reply:
Ahh, I see. I personally like the `Task` part to be a little more atomic, otherwise I feel like the state is spread over the `.state` property and all of the `.failed_results` links. I still think I'm looking at this from a totally different perspective of _not_ wanting retries (and other logic) to live in these types at all. 

In my view:
- A `Task` represents exactly one request, one network attempt, and zero or one responses.
- You can look at the `Task` itself to find out if it failed or succeeded
- if you want to retry a request, you'd spawn a new `Task`.

That feels more analogous to a `Promise` from JS (and I'd imagine a `Future` in async python), so perhaps we're just talking about different levels of abstraction where I'm seeing this as "queue some work for an event loop" instead of seeing this as something that has some baked-in domain logic.

At the moment, I think my own preference around behavior of these side-effecting functions in the `net` module is:

- One call = one attempt at delivery
- Every call returns some object that can be inspected about the state of the request, and maybe at some point later, the response

That means that it's up to the caller, so in the case of the `ext::auth` extension for instance, to know when and how to retry. In the future, we'd back up this logic with some kind of durable queue and domain-specific information about what kinds of requests are retryable under what conditions. For now, the auth extension could retry all network failures or 5XX responses in some fixed manner. But, I feel strongly that having any of that logic in this low-level module feels like it will cause us more trouble in the future. Imagine trying to implement a max-retry count in an extension, but because `net` always retries 5XX, you now how to observe both how many times you call `net::http::request` and the internal retry queue rather than knowing that every `net::http::request` call is exactly one outgoing request attempt. And, then you'd only be able to have a max_retries that is some multiple of the internal retry logic.

Reply:
Maybe `Task` -> `Future` or `AsyncResult`?

Reply:
I see your point now. It all makes sense to me! :+1:

Reply:
@fantix @1st1 

Any thoughts on the `Task` vs something else naming here? Does `Task` imply more capabilities than just "some work for an event loop to execute"?